### PR TITLE
[Conversion] convert triton ops(load2d/store2d/prefetch2d/dot) with block ptr to llvm

### PR DESCRIPTION
**Pull request by Dewei-Wang-sh**
```bash
ttgi.prefetch=> gen.matrix.2Dblockprefetch
tt.load => genx.matrix.2Dblockload
tt.store =>genx.matrix.2Dblockstore
tt.dot => gen.matrix.dpas
tt.make_tensor_ptr => v2i32(offsetx, offsety)
tt.advance => update v2i32(offsetx, offsety)
ttgi.glue => llvm.shuffle
ttgi.extract => llvm.shuflle
```

### CODE REVIEW

**Comment by Dewei-Wang-sh** [comment]
first round of rebase and format fix

File: include/triton/Dialect/TritonIntelGPU/IR/TritonIntelGPUOps.td [hunk 1]
```diff
@@ -89,10 +89,10 @@ def TTIG_ExtractOp : TTIG_Op<"extract", [Pure]> {
       - taking the 3rd subtensor (as indicated by the index) counting in
         column-major order
   }];
-  let arguments = (ins TT_TensorOrTensorPtr:$operand, I32Attr:$index);
+  let arguments = (ins TT_TensorOrTensorPtr:$base, I32Attr:$index);
```
Comment (Dewei-Wang-sh) [inline include/triton/Dialect/TritonIntelGPU/IR/TritonIntelGPUOps.td:92]: 'operand' is too ambiguous, extract op in other dialects use 'base/tensor/vector'

File: lib/Analysis/Utility.cpp [hunk 2]
```diff
@@ -911,7 +911,7 @@ MakeTensorPtrOp getMakeTensorPtrOp(Value v) {
   if (auto forOp = dyn_cast<scf::ForOp>(argOwner))
     return getMakeTensorPtrOp(
         forOp.getOperand(argNum + forOp.getNumControlOperands() - 1));
-  if (auto funcOp = dyn_cast<FuncOp>(argOwner)) {
+  if (auto funcOp = dyn_cast<FunctionOpInterface>(argOwner)) {
```
Comment (etiotto) [inline lib/Analysis/Utility.cpp:914]: Does this match what is upstream ? If not I would leave it as is to make merging easier.

File: lib/Conversion/TritonGPUToLLVM/TypeConverter.cpp [hunk 3]
```diff
@@ -17,12 +18,29 @@ TritonGPUToLLVMTypeConverter::TritonGPUToLLVMTypeConverter(
     MLIRContext *ctx, LowerToLLVMOptions &option,
     const DataLayoutAnalysis *analysis)
     : LLVMTypeConverter(ctx, option, analysis) {
-  addConversion([&](triton::PointerType type) -> std::optional<Type> {
-    return convertTritonPointerType(type);
-  });
-  addConversion([&](RankedTensorType type) -> std::optional<Type> {
-    return convertTritonTensorType(type);
-  });
+  if (mlir::triton::tools::getBoolEnv("INTEL_ENABLE_BLOCK_PTR")) {
```
Comment (etiotto) [inline lib/Conversion/TritonGPUToLLVM/TypeConverter.cpp:21]: Changing upstream file is to be avoided because it will make getting our implementation into the OpenAI github repository difficult. Is there a way to avoid it?

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp [hunk 4]
```diff
@@ -201,6 +203,12 @@ struct ConvertTritonGPUToLLVM
   void runOnOperation() override {
     MLIRContext *context = &getContext();
     ModuleOp mod = getOperation();
+    auto enableBlockPtr =
+        mlir::triton::tools::getBoolEnv("INTEL_ENABLE_BLOCK_PTR");
+    // fixme: set subgroupSize 16 for now
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp:208]: Use FIXME (uppercase  so we can grep more easily - other files use the uppercase form)

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp [hunk 5]
```diff
@@ -210,10 +218,12 @@ struct ConvertTritonGPUToLLVM
     int numCTAs = triton::gpu::TritonGPUDialect::getNumCTAs(mod);
     int threadsPerWarp = triton::gpu::TritonGPUDialect::getThreadsPerWarp(mod);
 
-    // Allocate shared memory and set barrier
-    ModuleAllocation allocation(mod);
-    ModuleMembarAnalysis membarPass(&allocation);
-    membarPass.run();
+    if (!enableBlockPtr) {
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp:221]: We have to consider the case the kernel doesn't use blocked pointers (or doesn't contain a dot operation).  IMO the way to do this is to have an analysis that identifies kernels that (a) use `tt.dot` and (b) use blocked pointers and only in that case avoid allocating shared memory ?

**Review by etiotto — DISMISSED** [review]
This PR contains several new new transformations and changes to the default processing (e.g. avoid allocation shared memory). This PR need to be split into several smaller self contained pieces. 

I'd like a separate PR for each new transformation. Each of these Prs will need to also add lit tests to test each of the transformation in isolation. Each transformation should also be properly documented, contain example of what it does, and in general the code need more comments to ease readability and understanding of the code.

File: lib/Analysis/Utility.cpp (hunk 2 above)
Comment (Dewei-Wang-sh) [reply lib/Analysis/Utility.cpp:914]: yes, do not break upstream.
I can upstream this change later.

File: lib/Conversion/TritonGPUToLLVM/TypeConverter.cpp (hunk 3 above)
Comment (Dewei-Wang-sh) [reply lib/Conversion/TritonGPUToLLVM/TypeConverter.cpp:21]: another way would be having our own copy of typeConverter.
but the drawback would be that it's hard to follow/leverage upstream changes in the future.
which do you prefer?

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp (hunk 4 above)
Comment (Dewei-Wang-sh) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp:208]: sure, also there is still many "auto", will change later.

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp (hunk 5 above)
Comment (Dewei-Wang-sh) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp:221]: yes, allocating shared memory is not conflict to using block pointer.
it's that the current case "gemm/attention" we are focusing have no shared memory involved.
without this condition branch, it fails in the membarpass,
and I do not want to spend time to dive into it, so mask it out works.
and later when supporting block ptr with shared memory, we can handle it.

**Comment by Dewei-Wang-sh** [comment]
> This PR contains several new new transformations and changes to the default processing (e.g. avoid allocation shared memory). This PR need to be split into several smaller self contained pieces.
> 
> I'd like a separate PR for each new transformation. Each of these Prs will need to also add lit tests to test each of the transformation in isolation. Each transformation should also be properly documented, contain example of what it does, and in general the code need more comments to ease readability and understanding of the code.

all the pieces here are need to make the one conversion work(-convert-tritongpu-to-llvm)
for example, if I don't use a env variable to control "shared mem allocation"  and "type converter" and "populate conversion passter", the pass would fail for a lit test.
but sure, will add lit test and comments to make it sound.

**Comment by Dewei-Wang-sh** [comment]
todo
1) change 'auto' to actual type
2) add lit test

File: lib/Conversion/TritonGPUToLLVM/TypeConverter.cpp (hunk 3 above)
Comment (etiotto) [reply lib/Conversion/TritonGPUToLLVM/TypeConverter.cpp:21]: Changing the upstream type converter goes against our long term goal of upstreaming the Intel GPU BE to OpenAI so I feel at this point is better to have our type converter.

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp (hunk 5 above)
Comment (etiotto) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp:222]: That increase the technical debt we carry so IMO we have to fix the problem from the start.

File: lib/Conversion/TritonGPUToLLVM/TypeConverter.cpp (hunk 3 above)
Comment (whitneywhtsang) [reply lib/Conversion/TritonGPUToLLVM/TypeConverter.cpp:21]: How about having our typeconverter inheriting from the upstream version?

Comment (etiotto) [reply lib/Conversion/TritonGPUToLLVM/TypeConverter.cpp:21]: Yes I think if we can extend the upstream one (e.g. inherit from it) and customize it with some of this code that would be preferable..

File: lib/Analysis/Utility.cpp (hunk 2 above)
Comment (etiotto) [reply lib/Analysis/Utility.cpp:914]: I think doing this upstream first also works. Is this really required for this feature ?

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp [hunk 6]
```diff
@@ -201,6 +203,12 @@ struct ConvertTritonGPUToLLVM
   void runOnOperation() override {
     MLIRContext *context = &getContext();
     ModuleOp mod = getOperation();
+    auto enableBlockPtr =
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp:206]: So the conversion pipeline need to be different for kernels that do use blocked pointers and kernels that do not.  I think we can wrap the logic in a `ConversionPipeline` class responsible for creating the actual pipeline.  That class can have a member function that analyzes the kernel to determine whether it uses blocked pointers or not, and use that analysis to determine what pipeline to use. 

Adding an env. variable to enable the "blocked pointer" pipeline would force us to run Triton tests with/without that option so that is a short term solution. We will need to have the analysis done asap.

File: include/triton/Dialect/TritonIntelGPU/IR/TritonIntelGPUOps.td (hunk 1 above)
Comment (etiotto) [inline include/triton/Dialect/TritonIntelGPU/IR/TritonIntelGPUOps.td:92]: Done in separate PR (https://github.com/intel/intel-xpu-backend-for-triton/pull/918)

File: lib/Dialect/TritonIntelGPU/IR/Ops.cpp [hunk 7]
```diff
@@ -147,7 +147,7 @@ LogicalResult GlueOp::verify() {
 
 LogicalResult ExtractOp::verify() {
   Type resultType = getRes().getType();
-  Type operandType = getOperand().getType();
+  Type operandType = getBase().getType();
```
Comment (etiotto) [inline lib/Dialect/TritonIntelGPU/IR/Ops.cpp:150]: Done in separate PR (https://github.com/intel/intel-xpu-backend-for-triton/pull/918)

File: lib/Analysis/Utility.cpp (hunk 2 above)
Comment (Dewei-Wang-sh) [reply lib/Analysis/Utility.cpp:914]: yes, because upstream only considered tt.func, 
in this case I use it in a func.func, so functionOpInterface works well for both cases.

Comment (Dewei-Wang-sh) [reply lib/Analysis/Utility.cpp:914]: upstream pr created
https://github.com/openai/triton/pull/3697

File: lib/Conversion/TritonGPUToLLVM/TypeConverter.cpp (hunk 3 above)
Comment (Dewei-Wang-sh) [reply lib/Conversion/TritonGPUToLLVM/TypeConverter.cpp:21]: got it, inhert is great

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp (hunk 6 above)
Comment (Dewei-Wang-sh) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp:206]: yes, env var is a short term solution

**Comment by Dewei-Wang-sh** [comment]
add lit test

**Comment by Dewei-Wang-sh** [comment]
> todo
> 
> 1. change 'auto' to actual type
> 2. add lit test

done

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp (hunk 6 above)
Comment (Dewei-Wang-sh) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp:206]: https://github.com/intel/intel-xpu-backend-for-triton/issues/913
in this pass, we plan to raise non-block ptr to block ptr as possible, and then determine which pipeline to go

File: include/triton/Tools/Sys/GetEnv.hpp [hunk 8]
```diff
@@ -44,6 +44,7 @@ inline const std::set<std::string> ENV_VARS = {
     "TRITON_ENABLE_LLVM_DEBUG",
     "USE_TTGIR_LOC",
     "TRITON_INTEL_EMULATE_FP16_ATOMICS",
+    "INTEL_ENABLE_BLOCK_PTR",
```
Comment (etiotto) [inline include/triton/Tools/Sys/GetEnv.hpp:47]: Added this in PR https://github.com/intel/intel-xpu-backend-for-triton/pull/939

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp [hunk 9]
```diff
@@ -17,6 +17,7 @@
 
 #include "intel/include/GPUToTritonGEN/GPUToTritonGENPass.h"
 #include "intel/include/TritonGENToLLVM/TritonGENToLLVMPass.h"
+#include "intel/include/TritonIntelGPUToLLVM/Passes.h"
```
[the remark this answers is not in the document]
Comment (Dewei-Wang-sh) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp:20]: that would be great!

File: third_party/intel/lib/TritonIntelGPUToLLVM/CMakeLists.txt [hunk 10]
```diff
@@ -1,26 +1,28 @@
 add_triton_library(TritonIntelGPUToLLVM
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/CMakeLists.txt:1]: Note: reorder in alphabetical order and added TypeConverter.cpp

**Comment by etiotto** [comment]
@whitneywhtsang please take another look.

**Comment by etiotto** [comment]
@Dewei-Wang-sh I am done with this one.  Take a peek and let me know if you spot some issues. Thanks.

**Review by Dewei-Wang-sh — COMMENTED** [review]
thanks for the effort, now it's in a very good shape.
lgtm

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+public:
+  using ConvertTritonGPUOpToLLVMPattern<
+      MakeTensorPtrOp>::ConvertTritonGPUOpToLLVMPattern;
+  LogicalResult
+  matchAndRewrite(MakeTensorPtrOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+
+    IntegerType i32Type = rewriter.getI32Type();
+    VectorType v2i32 = VectorType::get(2, i32Type);
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:39]:
```suggestion
    VectorType v2i32 = vec_ty(i32Type, 2);
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 12]
```diff
@@ -0,0 +1,404 @@
...
+    Value offsetX = op.getOffsets()[1];
+    Value offsetY = op.getOffsets()[0];
+    Value payLoad = rewriter.create<LLVM::UndefOp>(loc, v2i32);
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:42]:
```suggestion
    Value payLoad = undef(v2i32);
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 13]
```diff
@@ -0,0 +1,404 @@
...
+    Value idx0 = createIntConstant(i32Type, 0, rewriter, loc);
+    Value idx1 = createIntConstant(i32Type, 1, rewriter, loc);
+    payLoad =
+        rewriter.create<LLVM::InsertElementOp>(loc, payLoad, offsetX, idx0);
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:46]:
```suggestion
        insert_element(payLoad, offsetX, idx0);
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+    payLoad =
+        rewriter.create<LLVM::InsertElementOp>(loc, payLoad, offsetY, idx1);
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:48]:
```suggestion
        insert_element(payLoad, offsetY, idx1);
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+        if (auto attr = dyn_cast<mlir::IntegerAttr>(cst.getValue());
+            attr && attr.getInt() == 0)
+          continue;
+
+      IntegerType i32Type = rewriter.getI32Type();
+      Value idx = (i == 0)
+                      ? rewriter.create<LLVM::ConstantOp>(
+                            loc, i32Type, rewriter.getIntegerAttr(i32Type, 1))
+                      : rewriter.create<LLVM::ConstantOp>(
+                            loc, i32Type, rewriter.getIntegerAttr(i32Type, 0));
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:79]:
```suggestion
      Value idx = i32_val(!i);
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 16]
```diff
@@ -0,0 +1,404 @@
...
+      Value oldOffset = rewriter.create<LLVM::ExtractElementOp>(loc, ptr, idx);
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:80]:
```suggestion
      Value oldOffset = extract_element(ptr, idx);
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 17]
```diff
@@ -0,0 +1,404 @@
...
+      Value newOffset =
+          rewriter.create<LLVM::AddOp>(loc, i32Type, oldOffset, offset);
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:82]:
```suggestion
          add(i32Type, oldOffset, offset);
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 18]
```diff
@@ -0,0 +1,404 @@
...
+public:
+  using ConvertTritonGPUOpToLLVMPattern<
+      AdvanceOp>::ConvertTritonGPUOpToLLVMPattern;
+  LogicalResult
+  matchAndRewrite(AdvanceOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+    SmallVector<Value> offsets = adaptor.getOffsets();
+    Value ptr = adaptor.getPtr();
+    for (size_t i = 0; i < offsets.size(); i++) {
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:67]:
```suggestion
    for (size_t i = 0; i < offsets.size(); ++i) {
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 19]
```diff
@@ -0,0 +1,404 @@
...
+      IntegerType i32Type = rewriter.getI32Type();
+      Value idx = (i == 0)
+                      ? rewriter.create<LLVM::ConstantOp>(
+                            loc, i32Type, rewriter.getIntegerAttr(i32Type, 1))
+                      : rewriter.create<LLVM::ConstantOp>(
+                            loc, i32Type, rewriter.getIntegerAttr(i32Type, 0));
+      Value oldOffset = rewriter.create<LLVM::ExtractElementOp>(loc, ptr, idx);
+      Value newOffset =
+          rewriter.create<LLVM::AddOp>(loc, i32Type, oldOffset, offset);
+      ptr = rewriter.create<LLVM::InsertElementOp>(loc, ptr, newOffset, idx);
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:83]:
```suggestion
      ptr = insert_element(ptr, newOffset, idx);
```

**Review by etiotto — COMMENTED** [review]
Splitted the type converted related changes into #966

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 20]
```diff
@@ -0,0 +1,404 @@
...
+    if (auto cast =
+            dyn_cast<mlir::UnrealizedConversionCastOp>(base.getDefiningOp()))
+      base = cast.getInputs()[0];
+
+    OpBuilder::InsertPoint insertPoint = rewriter.saveInsertionPoint();
+    rewriter.setInsertionPointAfter(ptrOp);
+    Value bytes = createIntConstant(
+        i32Type, tensorType.getElementType().getIntOrFloatBitWidth() / 8,
+        rewriter, loc);
+    Value one = createIntConstant(i32Type, 1, rewriter, loc);
```
[the remark this answers is not in the document]
Comment (etiotto) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp]: done

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 21]
```diff
@@ -0,0 +1,366 @@
...
+           "only support 1d/2d load/store/prefetch for now");
+
+    Location loc = op.getLoc();
+    constexpr bool isLoad = std::is_same_v<OpType, LoadOp>;
+    constexpr bool isPrefetch = std::is_same_v<OpType, PrefetchOp>;
+
+    bool vnni = false, transpose = false;
+    if constexpr (isLoad) {
+      auto idxAttr = op->template getAttrOfType<mlir::IntegerAttr>("DotIdx");
+      vnni = idxAttr.getInt() == 1 ? true : false;
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:129]:
```suggestion
      vnni = idxAttr.getInt() == 1;
```

Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:129]: Should we also check dataSize < 32?

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 22]
```diff
@@ -0,0 +1,366 @@
...
+    std::iota(indices.begin(), indices.end(), start);
+    DenseI32ArrayAttr attr = rewriter.getDenseI32ArrayAttr(indices);
+    rewriter.replaceOpWithNewOp<LLVM::ShuffleVectorOp>(op, dstType, base, base,
+                                                       attr);
+    return success();
+  }
+};
+
+// FIXME: support it in upstream constantOpLowering
+class ArithConstantOpLowering
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:320]: Why it cannot be lowered by:
```
LogicalResult
ConstantOpLowering::matchAndRewrite(arith::ConstantOp op, OpAdaptor adaptor,
                                    ConversionPatternRewriter &rewriter) const {
  return LLVM::detail::oneToOneRewrite(op, LLVM::ConstantOp::getOperationName(),
                                       adaptor.getOperands(), op->getAttrs(),
                                       *getTypeConverter(), rewriter);
}
```
?

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp:1]: could we revert all changes to this file?

File: test/Conversion/tritongpu_to_llvm_intel_block_ptr.mlir [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+    %11 = arith.muli %8, %c256_i32 : i32
+    %12 = arith.muli %1, %c8_i32 : i32
+    %13 = arith.addi %12, %11 : i32
+    // CHECK:      [[undef:%.*]] = llvm.mlir.undef : vector<2xi32>
+    // CHECK-DAG:  [[zero:%.*]] = llvm.mlir.constant(0 : i32) : i32
+    // CHECK-NEXT: [[insert0:%.*]] = llvm.insertelement {{.*}}, [[undef]][[[zero]] : i32] : vector<2xi32>
+    // CHECK-NEXT:  [[one:%.*]] = llvm.mlir.constant(1 : i32) : i32
+    // CHECK-NEXT: [[insert1:%.*]] = llvm.insertelement {{.*}}, [[insert0]][[[one]] : i32] : vector<2xi32>
+    %14 = tt.make_tensor_ptr %arg0, [%c4096_i64, %c4096_i64], [%c4096_i64, %c1_i64], [%13, %c0_i32] {order = array<i32: 1, 0>} : <tensor<8x32xf16>, 1>
+    // CHECK: llvm.call @llvm.genx.GenISA.LSC2DBlockPrefetch.isVoid
```
Comment (whitneywhtsang) [inline test/Conversion/tritongpu_to_llvm_intel_block_ptr.mlir:42]: can we check the argument of the prefetch operation?

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp
Comment (etiotto) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp:1]: I want to remove include of "PatternTritonGPUOpToLLVM.h" because is not used.

Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TritonGPUToLLVM.cpp:1]: ok

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp (hunk 22 above)
Comment (etiotto) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:320]: I am going to remove the entire pattern, the unit test doesn't fail so ....  if we need to add it it can be done in a future PR.

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 24]
```diff
@@ -0,0 +1,324 @@
...
+  matchAndRewrite(DotOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto encodePrecision = [&](Type type) -> TritonGEN::PrecisionType {
+      if (type == bf16_ty)
+        return TritonGEN::PrecisionType::BF16;
+      else if (type == f16_ty)
+        return TritonGEN::PrecisionType::FP16;
+      else if (type == rewriter.getTF32Type())
+        return TritonGEN::PrecisionType::TF32;
+      assert(false && "add more support");
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:210]: llvm_unreachable?

**Review by whitneywhtsang — APPROVED** [review]

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp (hunk 22 above)
Comment (Dewei-Wang-sh) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp]: because we need to change it operands( resize in this case)
sure, we can add this later if not used by the current case.

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 25]
```diff
@@ -0,0 +1,404 @@
...
+                  ConversionPatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+
+    IntegerType i32Type = rewriter.getI32Type();
+    VectorType v2i32 = VectorType::get(2, i32Type);
+    Value offsetX = op.getOffsets()[1];
+    Value offsetY = op.getOffsets()[0];
+    Value payLoad = rewriter.create<LLVM::UndefOp>(loc, v2i32);
+    Value idx0 = createIntConstant(i32Type, 0, rewriter, loc);
+    Value idx1 = createIntConstant(i32Type, 1, rewriter, loc);
```
[the remark this answers is not in the document]
Comment (Dewei-Wang-sh) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp]: amazing, powerful macro!

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp (hunk 24 above)
Comment (Dewei-Wang-sh) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp:210]:
```suggestion
      llvm_unreachable("add more support for PrecisionType");
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp [hunk 26]
```diff
@@ -307,38 +307,6 @@ class ExtractOpConversion : public ConvertTritonGPUOpToLLVMPattern<ExtractOp> {
   }
 };
 
-// FIXME: support it in upstream constantOpLowering
-class ArithConstantOpLowering
```
Comment (Dewei-Wang-sh) [inline third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp]: ah, this shouldn't be removed, I need this.

Comment (etiotto) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp]: We decided during code review to remove it: https://github.com/intel/intel-xpu-backend-for-triton/pull/739#discussion_r1578681887

How is it needed ? Can you post a PR with a test to exercise the pattern ?

Comment (Dewei-Wang-sh) [reply third_party/intel/lib/TritonIntelGPUToLLVM/TritonOpsToLLVM.cpp]: oh, it's during my rebase for flash attention.
If I remember correctly, it's about attribute.
since I moved on quickly to resolve other rebase bugs, do not dive into it.
will exercise the pattern when post PRs for attention.
